README
Housekeeper, simple agent for password caching
==============================================

This project contains an agent that can be run to cache password of other
programs and a client library to use the agent.

It requires python-daemon and python-keyring to run.


Features
--------

* Minimal dependencies, only python 2.X and python-daemon
* Stores used passwords in memory, it doesn't use any keyring in filesystem
* Can be reset with SIGUSR1 or replaced with -r
* Stored keys have an expiration time, by default they expire in 10 minutes
* Can generate environment configuration easily sourceable from shell scripts
  or profile
* It's kept alive between sessions
* Implements backend for python keyring


Installation
-----------

If you got housekeeper as a source package, you can build it as egg or deb.

To build it as an egg, run:

  python setup.py bdist_egg

To build it as a debian package:

  dpkg-buildpackage


Usage
-----

Housekeeper can be used to temporary cache passowords you frequently use
in your sessions. For that, you need a running housekeeper daemon and your
application has to be able to talk with it. If you're interested in develop
an application that makes use of housekeeper, go to the next section.

`housekeeper` is the command used to start the daemon, run housekeeper -h to
see the complete list of available parameters. It always outputs the needed
environment configuration for your shell, so you can eval it, it also can
write with -e a file that you can source. The recommended usage is to run
it when starting your shell session, e.g. adding this lines to your .profile:

  if [[ -e $HOME/.housekeeper/env ]]; then
      . $HOME/.housekeeper/env
  else
      eval $(housekeeper --write-env-file --replace)
  fi

By default it writes the env file and other runtime information, as the
pidfile in $HOME/.housekeeper, but you can also use -d to specify another
directory.


Development
-----------

housekeeper basically implements a two command protocol to store and retrieve
keys from its internal cache in memory, the commands are:

  SET <service> <password> [<timeout>]

  GET <service>

SET is used to store a <password> for a <service> and GET to retrieve it, you
can specify a <timeout> in seconds when storing the password to make it expire
after so many seconds without being used. Only a 1 minute granularity is
guaranteed.

To connect with the daemon you have to use the socket file shown in the output
of the daemon, this file is in the running directory and is called `socket`, by
default: $HOME/.housekeeper/socket, a client application should use the
`HOUSEKEEPER_SOCKET` environment variable to find this socket file.

housekeeper also provides a client-side python API with two convenience methods
mapped to those commands, you can find an example of the use of this API in
examples/test_client.py, but basically:

  from housekeeper.client import HousekeeperClient

  client = HousekeeperClient()
  client.set('mercurial', 'foobar', timeout=3600)
  print client.get('mercurial')

It also provides a backend for python keyring module, to use it:

  import keyring

  from housekeeper.client import HousekeeperKeyringBackend

  housekeeper = HousekeeperKeyringBackend(timeout=3600)
  keyring.set_keyring(housekeeper)

Timeout is always optional and it's always defaulted to 600 seconds, if you set
it to 0, the key never expires.


debian/changelog
housekeeper (1.1-4) unstable; urgency=medium

  * Passwords are stored in base64 to avoid problems with special characters
  * Type errors on input parsing don't kill the daemon

 -- Jaime Soriano <jsoriano@tuenti.com>  Mon, 27 Aug 2012 18:31:53 +0200

housekeeper (1.1-3) unstable; urgency=low

  * Fixed infinite timeout for cache expiration

 -- Jaime Soriano <jsoriano@tuenti.com>  Tue, 21 Aug 2012 20:37:33 +0200

housekeeper (1.1-2) unstable; urgency=high

  * Security update, only owner of the process can write pid and env files

 -- Jaime Soriano <jsoriano@tuenti.com>  Thu, 09 Aug 2012 12:12:09 +0200

housekeeper (1.1-1) unstable; urgency=low

  * python-setuptools added as dependency

 -- JaimeSoriano <jsoriano@tuenti.com>  Mon, 04 Jun 2012 13:21:07 +0200

housekeeper (1.1) unstable; urgency=low

  * Version 1.1

 -- Jaime Soriano <jsoriano@tuenti.com>  Fri, 01 Jun 2012 16:31:09 +0200

housekeeper (1.0) unstable; urgency=low

  * Version 1.0

 -- Jaime Soriano <jsoriano@tuenti.com>  Fri, 01 Jun 2012 12:30:41 +0200

housekeeper (0.1dev-1) unstable; urgency=low

  * source package automatically created by stdeb 0.6.0

 -- Jaime Soriano <jsoriano@tuenti.com>  Fri, 01 Jun 2012 12:05:23 +0200


test_client.py
import time

from housekeeper import HousekeeperClient

c = HousekeeperClient()
c.set('mercurial', 'fooobar')
print c.get('mercurial')
c.set('mercurial', 'stuff', 1)
print c.get('mercurial')
time.sleep(3)
print c.get('mercurial')


examples/test_client.py
#!/usr/bin/python
# Copyright (c) 2011 Tuenti Technologies
# See LICENSE for details

import time

from housekeeper.client import HousekeeperClient

c = HousekeeperClient()
c.set('mercurial', 'fooobar')
print c.get('mercurial')
c.set('mercurial', 'stuff', 1)
print c.get('mercurial')
time.sleep(3)
print c.get('mercurial')


housekeeper/client.py
#!/usr/bin/python
# Copyright (c) 2011 Tuenti Technologies
# See LICENSE for details

import base64
import os
import os.path
import socket

from keyring.backend import KeyringBackend

#TODO: Better exceptions
class HousekeeperClientException(Exception): pass

if os.environ.has_key('HOUSEKEEPER_SOCKET'):
    default_socket = os.environ['HOUSEKEEPER_SOCKET']
else:
    default_socket = os.path.expanduser('~/.housekeeper/socket')

class HousekeeperClient:
    def __init__(self, socketfile=default_socket):
        self.socketfile = socketfile

    def __request(self, cmd):
        s = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        s.connect(self.socketfile)
        f = s.makefile('wr')
        f.write(cmd + '\r\n')
        f.flush()
        ret = f.read().strip().split('\r\n')

        if len(ret) == 1:
            if ret[0] == 'OK':
                return
            else:
                raise HousekeeperClientException
        else:
            response, result_code = ret
            s.close()
            if result_code == 'OK':
                return response
            else:
                raise HousekeeperClientException(response)

    def set(self, service, password, timeout=-1):
        cmd = "SET %s %s" % (service, base64.b64encode(password))
        if timeout >= 0:
            cmd += " %d" % timeout
        self.__request(cmd)

    def get(self, service):
        password = self.__request("GET %s" % service)
        try:
            return base64.b64decode(password)
        except TypeError:
            # Old running instances of housekeeper can have unencoded passwords
            # TODO: Improve protocol for 2.0, and remove this
            return password

class HousekeeperKeyringBackend(KeyringBackend):
    def __init__(self, socketfile=default_socket, timeout=600):
        self.socketfile = socketfile
        self.client = HousekeeperClient(socketfile)
        self.timeout = timeout

    def supported(self):
        if os.path.exists(self.socketfile):
            return 1
        else:
            return -1

    def get_password(self, service, username):
        try:
            return self.client.get("%s@%s" % (username, service))
        except HousekeeperClientException:
            return ''

    def set_password(self, service, username, password):
        self.client.set("%s@%s" % (username, service), password, self.timeout)
        return 0


housekeeper/keyring.py
#!/usr/bin/python
# Copyright (c) 2012 Tuenti Technologies
# See LICENSE for details

import os
import os.path
import signal
import sys
import daemon
try:
    import daemon.pidlockfile as pidlockfile
except:
    import daemon.pidfile as pidlockfile
import threading
import time
import socket

class MissingKeyException(Exception): pass
class InvalidKeyException(Exception): pass
class InvalidCommandException(Exception): pass


class HousekeeperKeyring:
    def __init__(self):
        self.keys_lock = threading.Lock()
        self.reset()

    def reset(self):
        self.keys_lock.acquire()
        self.keys = {}
        self.keys_lock.release()

    def check_expiration(self, service=None):
        def check(key):
            if key['expiration'] and key['expiration'] < time.time():
                del self.keys[service]

        self.keys_lock.acquire()
        if service:
            if service in self.keys:
                check(self.keys[service])
        else:
            services = self.keys.keys()
            for service in services:
                check(self.keys[service])
        self.keys_lock.release()

    def set_key(self, service, key, timeout=600):
        if not service or not key or timeout < 0:
            raise InvalidKeyException()

        self.keys_lock.acquire()
        self.keys[service] = {
            'key': key,
            'expiration': time.time() + timeout,
            'timeout': timeout,
        }
        self.keys_lock.release()

    def get_key(self, service):
        self.check_expiration(service)
        self.keys_lock.acquire()
        try:
            if self.keys.has_key(service):
                key = self.keys[service]
                key['expiration'] = time.time() + key['timeout']
                result = key['key']
            else:
                raise MissingKeyException("No key for '%s' service" % service)
        finally:
            self.keys_lock.release()
        return result
        

class HousekeeperDaemon(daemon.DaemonContext):
    def __init__(self, directory=os.path.expanduser('~/.housekeeper'), write_envfile=False, replace=True):
        self.directory = directory
        self.pidfile = os.path.join(directory, 'pid')
        self.envfile = os.path.join(directory, 'env')
        self.socketfile = os.path.join(directory, 'socket')
        self.keyring = HousekeeperKeyring()
        self.write_envfile = write_envfile
        self.replace = replace

        self.__startup()

        if not os.path.exists(self.directory):
            os.makedirs(self.directory)

        self.env_content = "HOUSEKEEPER_SOCKET='%s'; export HOUSEKEEPER_SOCKET;" % self.socketfile
        print self.env_content
        sys.stdout.flush()

        super(HousekeeperDaemon, self).__init__(
            working_directory = self.directory,
            pidfile = pidlockfile.PIDLockFile(self.pidfile),
            detach_process = True,
        )

        self.signal_map = {
            signal.SIGTERM: 'terminate',
            signal.SIGHUP: None,
            signal.SIGUSR1: self.reset_keyring,
        }
        
    def __startup(self):
        # Check if daemon already running
        if os.path.isfile(self.pidfile):
            try:
                pid = int(open(self.pidfile).read())
                # Check For the existence of a unix pid.
                os.kill(pid, 0)
            except (OSError, ValueError):
                sys.stderr.write("Removing stale locking file: %s\n" % self.pidfile)
                os.unlink(self.pidfile)
            else:
                if self.replace:
                    sys.stderr.write("Replacing daemon PID: %d\n" % pid)
                    os.kill(pid, signal.SIGTERM)
                else:
                    sys.stderr.write("Daemon already running PID: %d\n" % pid)
                    exit(1)

    def __check_expiration(self):
        class Checker(threading.Thread):
            def __init__(self, keyring):
                super(Checker, self).__init__()
                self.keyring = keyring
                self.event = threading.Event()
                # TODO: Kill the thread with stop() instead of with self.daemon
                self.daemon = True

            def run(self):
                while not self.event.isSet():
                    self.keyring.check_expiration()
                    self.event.wait(60)

            def stop(self):
                self.event.set()

        return Checker(self.keyring)

    def reset_keyring(self):
        self.keyring.reset()

    def handle_command(self, command):
        tokens = command.split(' ')
        if tokens[0] == 'GET' and len(tokens) == 2:
            return self.keyring.get_key(tokens[1])
        elif tokens[0] == 'SET' and len(tokens) >= 3:
            args = [tokens[1], tokens[2]]
            if len(tokens) > 3:
                args.append(int(tokens[3]))
            self.keyring.set_key(*args)
            return ''
        else:
            raise InvalidCommandException()

    def handle_connection(self, conn):
        socket_file = conn.makefile('rw')
        try:
            command = socket_file.readline()
            response = self.handle_command(command.strip())
            if response:
                socket_file.write(response + '\r\n')
            socket_file.write('OK\r\n')
            socket_file.flush()
            socket_file.close()
        except MissingKeyException:
            socket_file.write('MissingKeyException\r\nERROR\r\n')
        except InvalidKeyException:
            socket_file.write('InvalidKeyException\r\nERROR\r\n')
        except InvalidCommandException:
            socket_file.write('InvalidCommandException\r\nERROR\r\n')
        finally:
            conn.close()

    def close(self):
        if os.path.exists(self.socketfile):
            os.unlink(self.socketfile)
        if os.path.exists(self.envfile):
            os.unlink(self.envfile)
        super(HousekeeperDaemon, self).close()

    def main(self):
        self.open()

        self.checker = self.__check_expiration()
        self.checker.start()

        if os.path.lexists(self.socketfile):
            os.unlink(self.socketfile)

        if self.write_envfile:
            env = open(self.envfile, 'w')
            env.write(self.env_content)
            env.close()

        self.socket = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        umask = os.umask(0077)
        self.socket.bind(self.socketfile)
        umask = os.umask(umask)
        self.socket.listen(1)

        while True:
            conn, addr = self.socket.accept()
            self.handle_connection(conn)



def main():
    from optparse import OptionParser

    parser = OptionParser()
    parser.add_option("-d", "--directory",
        dest="directory",
        help="Uses DIRECTORY as working directory",
        metavar="DIRECTORY",
        default=os.path.expanduser('~/.housekeeper'),
    )
    parser.add_option("-e", "--write-env-file",
        dest="write_envfile",
        help="Write environment file in working directory",
        action="store_true",
        default=False,
    )
    parser.add_option("-r", "--replace",
        dest="replace",
        help="Replace running daemon if any",
        action="store_true",
        default=False,
    )

    (options, args) = parser.parse_args()

    daemon = HousekeeperDaemon(
        directory=options.directory,
        write_envfile=options.write_envfile,
        replace=options.replace,
    )
    daemon.main()


if __name__ == "__main__":
    main()



housekeeper/githelper.py
#!/usr/bin/python
# Copyright (c) 2012 Tuenti Technologies
# See LICENSE for details

import os
import os.path
import sys

from client import HousekeeperClient, HousekeeperClientException

def get_request():
    request = {}
    while True:
        line = sys.stdin.readline()
        if not line.strip():
            break
        key, value = line.strip().split('=')
        request[key] = value
    return request

def put_response(response):
    for key, value in response.iteritems():
        print "%s=%s" % (key, value)

def main():
    from optparse import OptionParser

    usage = "usage: %prog [options] <action>"
    parser = OptionParser(usage=usage)
    parser.add_option("-d", "--directory",
        dest="directory",
        help="Uses DIRECTORY as working directory",
        metavar="DIRECTORY",
        default=os.path.expanduser('~/.housekeeper'),
    )
    parser.add_option("-t", "--timeout",
        dest="timeout",
        help="Uses TIMEOUT seconds as timeout (Default: 900)",
        metavar="DIRECTORY",
        type="int",
        default=900,
    )

    (options, args) = parser.parse_args()

    if len(args) != 1:
        parser.print_help()
        return

    socketfile = os.path.join(options.directory, 'socket')
    action = args[0].strip()

    if not action in ['get', 'set']: return

    request = get_request()
    client = HousekeeperClient(socketfile)

    if not 'username' in request: return

    service = "%s@git" % (request['username'])
    if action == 'get':
        response = request
        try:
            response['password'] = client.get(service)
            put_response(response)
        except HousekeeperClientException:
            pass
    elif action == 'set':
        if not 'password' in request: return
        client.set(service, request['password'], options.timeout)

if __name__ == "__main__":
    main()


housekeeper/housekeeper.py
#!/usr/bin/python
# Copyright (c) 2012 Tuenti Technologies
# See LICENSE for details

import os
import os.path
import signal
import sys
import daemon
try:
    import daemon.pidlockfile as pidlockfile
except:
    import daemon.pidfile as pidlockfile
import threading
import time
import socket

class MissingKeyException(Exception): pass
class InvalidKeyException(Exception): pass
class InvalidCommandException(Exception): pass


class HousekeeperKeyring:
    def __init__(self):
        self.keys_lock = threading.Lock()
        self.reset()

    def reset(self):
        self.keys_lock.acquire()
        self.keys = {}
        self.keys_lock.release()

    def check_expiration(self, service=None):
        def check(key):
            if key['timeout'] and key['expiration'] < time.time():
                del self.keys[service]

        self.keys_lock.acquire()
        if service:
            if service in self.keys:
                check(self.keys[service])
        else:
            services = self.keys.keys()
            for service in services:
                check(self.keys[service])
        self.keys_lock.release()

    def set_key(self, service, key, timeout=600):
        if not service or not key or timeout < 0:
            raise InvalidKeyException()

        self.keys_lock.acquire()
        self.keys[service] = {
            'key': key,
            'expiration': time.time() + timeout,
            'timeout': timeout,
        }
        self.keys_lock.release()

    def get_key(self, service):
        self.check_expiration(service)
        self.keys_lock.acquire()
        try:
            if self.keys.has_key(service):
                key = self.keys[service]
                key['expiration'] = time.time() + key['timeout']
                result = key['key']
            else:
                raise MissingKeyException("No key for '%s' service" % service)
        finally:
            self.keys_lock.release()
        return result


class HousekeeperDaemon(daemon.DaemonContext):
    def __init__(self, directory=os.path.expanduser('~/.housekeeper'), write_envfile=False, replace=True):
        self.directory = directory
        self.pidfile = os.path.join(directory, 'pid')
        self.envfile = os.path.join(directory, 'env')
        self.socketfile = os.path.join(directory, 'socket')
        self.keyring = HousekeeperKeyring()
        self.write_envfile = write_envfile
        self.replace = replace

        self.__startup()

        if not os.path.exists(self.directory):
            os.makedirs(self.directory)

        self.env_content = "HOUSEKEEPER_SOCKET='%s'; export HOUSEKEEPER_SOCKET;" % self.socketfile
        print self.env_content
        sys.stdout.flush()

        super(HousekeeperDaemon, self).__init__(
            working_directory = self.directory,
            pidfile = pidlockfile.PIDLockFile(self.pidfile),
            detach_process = True,
        )

        self.signal_map = {
            signal.SIGTERM: 'terminate',
            signal.SIGHUP: None,
            signal.SIGUSR1: lambda *args: self.reset_keyring(),
        }

    def __startup(self):
        # Check if daemon already running
        if os.path.isfile(self.pidfile):
            try:
                pid = int(open(self.pidfile).read())
                # Check For the existence of a unix pid.
                os.kill(pid, 0)
            except (OSError, ValueError):
                sys.stderr.write("Removing stale locking file: %s\n" % self.pidfile)
                os.unlink(self.pidfile)
            else:
                if self.replace:
                    sys.stderr.write("Replacing daemon PID: %d\n" % pid)
                    os.kill(pid, signal.SIGTERM)
                else:
                    sys.stderr.write("Daemon already running PID: %d\n" % pid)
                    exit(1)

    def __check_expiration(self):
        class Checker(threading.Thread):
            def __init__(self, keyring):
                super(Checker, self).__init__()
                self.keyring = keyring
                self.event = threading.Event()
                # TODO: Kill the thread with stop() instead of with self.daemon
                self.daemon = True

            def run(self):
                while not self.event.isSet():
                    self.keyring.check_expiration()
                    self.event.wait(60)

            def stop(self):
                self.event.set()

        return Checker(self.keyring)

    def reset_keyring(self):
        self.keyring.reset()

    def handle_command(self, command):
        tokens = command.split(' ')
        if tokens[0] == 'GET' and len(tokens) == 2:
            return self.keyring.get_key(tokens[1])
        elif tokens[0] == 'SET' and len(tokens) >= 3:
            args = [tokens[1], tokens[2]]
            if len(tokens) > 3:
                args.append(int(tokens[3]))
            self.keyring.set_key(*args)
            return ''
        else:
            raise InvalidCommandException()

    def handle_connection(self, conn):
        socket_file = conn.makefile('rw')
        try:
            command = socket_file.readline()
            response = self.handle_command(command.strip())
            if response:
                socket_file.write(response + '\r\n')
            socket_file.write('OK\r\n')
            socket_file.flush()
            socket_file.close()
        except MissingKeyException:
            socket_file.write('MissingKeyException\r\nERROR\r\n')
        except InvalidKeyException:
            socket_file.write('InvalidKeyException\r\nERROR\r\n')
        except InvalidCommandException:
            socket_file.write('InvalidCommandException\r\nERROR\r\n')
        except ValueError:
            socket_file.write('InvalidCommandException\r\nERROR\r\n')
        finally:
            conn.close()

    def close(self):
        if os.path.exists(self.socketfile):
            os.unlink(self.socketfile)
        if os.path.exists(self.envfile):
            os.unlink(self.envfile)
        super(HousekeeperDaemon, self).close()

    def main(self):
        self.open()
        umask = os.umask(0022)

        self.checker = self.__check_expiration()
        self.checker.start()

        if os.path.lexists(self.socketfile):
            os.unlink(self.socketfile)

        if self.write_envfile:
            env = open(self.envfile, 'w')
            env.write(self.env_content)
            env.close()

        self.socket = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        os.umask(0077)
        self.socket.bind(self.socketfile)
        os.umask(umask)
        self.socket.listen(1)

        while True:
            try:
                conn, addr = self.socket.accept()
                self.handle_connection(conn)
            except socket.error:
                pass


def main():
    from optparse import OptionParser

    parser = OptionParser()
    parser.add_option("-d", "--directory",
        dest="directory",
        help="Uses DIRECTORY as working directory",
        metavar="DIRECTORY",
        default=os.path.expanduser('~/.housekeeper'),
    )
    parser.add_option("-e", "--write-env-file",
        dest="write_envfile",
        help="Write environment file in working directory",
        action="store_true",
        default=False,
    )
    parser.add_option("-r", "--replace",
        dest="replace",
        help="Replace running daemon if any",
        action="store_true",
        default=False,
    )

    (options, args) = parser.parse_args()

    daemon = HousekeeperDaemon(
        directory=options.directory,
        write_envfile=options.write_envfile,
        replace=options.replace,
    )
    daemon.main()


if __name__ == "__main__":
    main()

